include/revng/Pipeline/ContainerSet.h

```diff
@@ -0,0 +1,143 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <cstddef>
+#include <fstream>
+#include <memory>
+#include <optional>
+#include <set>
+#include <sstream>
+#include <string>
+
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/MemoryBuffer.h"
+
+#include "revng/Pipeline/Container.h"
+#include "revng/Pipeline/ContainerFactory.h"
+#include "revng/Pipeline/Contract.h"
+#include "revng/Pipeline/Target.h"
+#include "revng/Support/Assert.h"
+
+namespace Pipeline {
+
+/// A map from containerName to a MaybeContainer, which is initially empty.
```

Comment:
1. You should mention this class ha the capability of magically instantiate containers through provided factories.
2. What's a `MaybeContainer`?
3. This can be very very expensive to copy, is that right? That's why the copy constructor is deleted and we have an explicit `clone` method. Nice! A bit surprising to see it being returned by copy so often despite this, but sounds like a nice design!

Reply:
yeah, the copy method would be fairly strange here,